src/test/java/htsjdk/samtools/CRAMSliceMD5Test.java

```diff
@@ -0,0 +1,105 @@
+        // check the mixed case source does not change ref base casing:
+        Assert.assertEquals(refBasesFromMixedCaseSource, referenceBases);
+        // check the mixed case source contains lower case bases:
+        Assert.assertTrue(new String(refBasesFromMixedCaseSource).contains("a"));
+
+        final int readLen = referenceBases.length;
+        final byte[] bases = new byte[readLen];
+        Arrays.fill(bases, (byte) 'A');
+        final byte[] scores = new byte[readLen];
+        Arrays.fill(scores, (byte) '!');
+
+        final SAMRecord record = new SAMRecord(samFileHeader);
+        record.setReadName("test");
+        record.setReferenceIndex(0);
+        record.setAlignmentStart(1);
+        record.setCigarString(readLen + "M");
+        record.setReadBases(bases);
+        record.setBaseQualities(scores);
+
+        // write a valid CRAM with a valid reference source:
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        final CRAMFileWriter writer = new CRAMFileWriter(baos, referenceSourceUpperCased, samFileHeader, "test");
+        writer.addAlignment(record);
+        writer.close();
+
+        // read the CRAM:
+        final ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
+        final CramHeader cramHeader = CramIO.readCramHeader(bais);
+        final Container container = ContainerIO.readContainer(cramHeader.getVersion(), bais);
+        final Slice slice = container.slices[0];
+        Assert.assertEquals(slice.alignmentStart, 1);
+        Assert.assertEquals(slice.alignmentSpan, referenceBases.length);
+        // check the slice MD5 is the MD5 of upper-cased ref bases:
+        byte[] ucRefMD5 = SequenceUtil.calculateMD5(refBasesFromUCSource, 0, refBasesFromUCSource.length);
+        Assert.assertEquals(slice.refMD5, ucRefMD5);
+
+        // check the CRAM file reads:
+        CRAMFileReader reader = new CRAMFileReader(new ByteArrayInputStream(baos.toByteArray()), (File) null, referenceSourceUpperCased, ValidationStringency.STRICT);
+        SAMRecordIterator iterator = reader.getIterator();
+        Assert.assertTrue(iterator.hasNext());
+        Assert.assertEquals(iterator.next(), record);
+
+        // try reading the CRAM file with the incorrect ref source that does not upper case bases:
+        reader = new CRAMFileReader(new ByteArrayInputStream(baos.toByteArray()), (File) null, referenceSourceMixedCase, ValidationStringency.STRICT);
+        iterator = reader.getIterator();
+        try {
+            // expect an exception here due to slice MD5 mismath:
+            Assert.assertTrue(iterator.hasNext());
+            Assert.fail("Reference MD5s must mismatch.");
+        } catch (final CRAMException e) {
+            Assert.assertTrue(e.getMessage().startsWith("Reference sequence MD5 mismatch for slice: sequence id 0, start 1, span 31, expected MD5 "));
+        }
+    }
+}
```

Comment:
There are two completely separate tests mixed together in this method; one that verifies the that default ReferenceSource implementation results in a correct slice MD5, and one that verifies that a bogus ReferenceSource implementation causes the slice MD5 calculation to fail. They should be factored into two separate tests with whatever common code is possible.